src/details/ArborX_DetailsEulerTour.hpp

```diff
@@ -0,0 +1,279 @@
+#include <ArborX_MinimumSpanningTree.hpp> // WeightedEdge
+
+namespace ArborX::Details
+{
+
+// Assumption: edges are bidirectional and unique and form a tree
+template <typename ExecutionSpace, typename Edges>
+Kokkos::View<int *, typename Edges::memory_space>
+eulerTourList(ExecutionSpace const &exec_space, Edges const &edges)
+{
+  Kokkos::Profiling::pushRegion("ArborX::euler_tour_list");
+
+  using MemorySpace = typename Edges::memory_space;
+
+  int const num_edges = edges.size();
+  int const num_vertices = num_edges + 1;
+
+  ARBORX_ASSERT(num_edges > 0);
+
+  Kokkos::View<int *, MemorySpace> offsets("ArborX::euler_tour_list::offsets",
+                                           num_vertices + 1);
+  Kokkos::parallel_for(
+      "ArborX::euler_tour_list::compute_counts",
+      Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, num_edges),
+      KOKKOS_LAMBDA(int const edge_index) {
+        auto const &edge = edges(edge_index);
+        Kokkos::atomic_increment(&offsets(edge.source));
+        Kokkos::atomic_increment(&offsets(edge.target));
+      });
+  exclusivePrefixSum(exec_space, offsets);
+
+  ARBORX_ASSERT(KokkosExt::lastElement(exec_space, offsets) == 2 * num_edges);
+
+  Kokkos::View<unsigned *, MemorySpace> permute(
+      Kokkos::view_alloc(exec_space, Kokkos::WithoutInitializing,
+                         "ArborX::euler_tour_list::permute"),
+      2 * num_edges);
+
+  auto offsets_clone = KokkosExt::clone(exec_space, offsets);
+  Kokkos::parallel_for(
+      "ArborX::euler_tour_list::compute_intermediate_permutation",
+      Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, num_edges),
+      KOKKOS_LAMBDA(int const edge_index) {
+        auto const &edge = edges(edge_index);
+        permute(Kokkos::atomic_fetch_add(&offsets_clone(edge.source), 1)) =
+            2 * edge_index + 0;
+        permute(Kokkos::atomic_fetch_add(&offsets_clone(edge.target), 1)) =
+            2 * edge_index + 1;
+      });
+  Kokkos::resize(offsets_clone, 0); // deallocate memory
+
+  Kokkos::View<int *, MemorySpace> successors(
+      Kokkos::view_alloc(exec_space, Kokkos::WithoutInitializing,
+                         "ArborX::euler_tour_list::successors"),
+      2 * num_edges);
+  Kokkos::parallel_for(
+      "ArborX::euler_tour_list::build_successors",
+      Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, num_vertices),
+      KOKKOS_LAMBDA(int const i) {
+        int start = offsets(i);
+        int len = offsets(i + 1) - start;
+
+        auto value = [i, start, &edges, &permute](int k) {
+          auto const &edge = edges(permute(start + k) / 2);
+          return (i == edge.source ? edge.target : edge.source);
+        };
+
+        // FIXME This sorting is unnecessary. It will only affect the order of
+        // subtrees to traverse, but will still produce a correct Euler Tour.
+        // However, removing the sort will make the algorithm
+        // non-deterministic, so the testing would need to change.
+        //
```

Comment:
So I'd like to remove the sort. But need to figure out how to test the successors list as it will be undeterministic.